### ISSUE TITLE
Need to fix Id seed and IDPrefix

### ISSUE THREAD

**Issue by oznetmaster** [nunit/nunit#2541]
As per my comments in #2529, there are a number of issues with both ITest.Id and Test.IdPrefix.

These include:

The seed for Test.Id is statically initialized, so that if the test assembly is never unloaded, the value is never reset to the default seed.

There is no way set IdPrefix in nunit-lite.  Currently, it is only set in FrameworkController, which is not used by nunit-lite.

If a test instantiates a new test tree by instantiating a new FrameworkController, the current IdPrefix is permanently overwritten.

There is currently no way set the Id seed programatically, as there is with IdPrefix.  This makes it impossible to validate that identical test trees are generated from the same test source.

These all appear to be easy fixes.  They will require additions to the existing api.  For instance, an optional Id seed in the FrameworkController constructor, and optional IdPrefix and Id seed support in the nunit-lite command line and runner.  These would be fully backward compatible.  The current IdPrefix and Id might need to be stored in the TestExcutionContext for nesting, but that is internal to nunit.

**Comment by oznetmaster** [comment]
These changes will also allow a set of tests to be written that test Id generation and Id nesting, which it is currently difficult to do, as noted by @CharliePoole  in #2529.

**Comment by oznetmaster** [comment]
It also seems that the setters for Id, IdPrefix and IdSeed (new) should all be internal.

**Comment by CharliePoole** [comment]
This all looks fixable, but I think we need to define more clearly the use case we are trying to support. As I noted on #2529 I have been unable to replicate the problem either running executable tests directly or by use of the nunitlite-runner from the command-line. I'm guessing that you are doing something else, perhaps running nunitlite programmatically?

A second underlying assumption was that runners other than nunitlite would use the framework API, that is the reflection API supported by the framework controller.

As used by the console and gui runners, a unique prefix is provided by each instance of the framework driver. That is what permits the use of the static seed value for the id. Each ApplicationDomain generates a new series of ids but applies the prefix provided by the driver in order to insure uniqueness.

For the same reason, second-level tests don't present a problem because they are run in a separate ApplicationDomain.

All of this merely goes to point out that your specific usage needs to be taken into account and then built into the engine. I definitely agree we should spell out some clear expectations around testid and write a set of tests to verify them.

**Comment by CharliePoole** [comment]
I agree regarding internal. See issue #1426

**Comment by oznetmaster** [comment]
I only run nunit-lite from the command line.

The code is pretty clear, as I posted in #2529.  The Id seed is a static variable inside of Test, and is set to 1000 when the assembly is loaded.  It is NEVER modified except to add 1 to it.

So the only way it is reset to 1000 is if it is reloaded.

I assume that in most environments (those that support ApplicationDomains), each explore or run operation loads the test assembly into a new ApplicationDomain for that operation?

However, there are environments (such a my current CF one, and the new (special) mono environment that I am working with) that do not support ApplicationDomains.  The test assembly is only loaded once, and remains in memory for all subsequent explore or test runs.

In that environment, the Id seed (_nextId) is never reset, and keeps counting up forever.  Also, any change to IdPrefix stays forever until it is changed again.

I think that a design which makes the assumption that the test assembly is reloaded for every explore or run operation is faulty at the design level.  Yes, that does "force" re-initialization of all static variables, but it hides a possible multitude of problems.  It also is almost untestable.

It is very simple to remove these problems.  No more then about ten lines of code added or modified.

I do not know if there will be more environments in the future with a single ApplicationDomain, but I have one old one, and one new one, so, for me, that is an indication that the problem should be rectified, given how simple the solutions is.

It does raise the question of other ApplicationDomain related dependencies, but I have not seen any other to date.  That does not mean they do not exist.  Looking at other statically initialized member variables might point others.

**Comment by oznetmaster** [comment]
In my cases, there is no engine, because these environments do not support remoting.  Only nunit-lite support is really an issue for me.

When the engine stops using remoting, this could change.

**Comment by oznetmaster** [comment]
I will point out that I am already doing these fixes for my version of NUnit, because I have to.  They are being made as described previousuly.

I am happy to make them available to the release if they are thought useful.

**Comment by CharliePoole** [comment]
I have to take a small exception to what you write. Upon creation of the code, neither I nor anyone else working on it would have envisioned an environment in whch multiple command-line executions would __not__ re-initialize statics. I'm stunned to learn that's what happens and it seems to me to be a failure at the C# level. Nevertheless, I agree we should deal with it.

Can you take this one on via a PR? 

One thing strikes me... any initialization we are doing in the FrameworkManager class should probably be moved to the corresponding method of the runner that is being called unless we __only__ want the initialization to occur when using that API.

**Comment by oznetmaster** [comment]
That is what I planned to do.

**Comment by oznetmaster** [comment]
I think it is because if the specific environment in which I am running and not C# itself.

With a single ApplicationDomain, and no command line processor in the system, NUnit-lite is loaded as a program, and then waits for console input, which it then parses.  It will accept multiple command lines without being unloaded, just as if it were a command line processor.  However, it initializes each static variable only once, and each static constructor only runs once, no matter how many command lines are entered, because that is how a C# program behaves.

I could restart nunit-lite for each command lines, but the cost of doing that is significant - about 30 seconds to a minute per restart.  That would solve the problem, however.

I would rather fix it by dynamically initializing each affected static variable at the start of each operation.  That is a minor overhead compared to a full program restart.

**Comment by CharliePoole** [comment]
That explains a lot. My way of testing, when I worked on CF, reloaded the app each time.

"While you are at it" you might want to look at the initialization of the random seed, which involves a static.

**Comment by oznetmaster** [comment]
In fact, as far as I can tell, the default InitialSeed value which is set in the static constructor for Randomizer is the only other suspect.

However, it is set to new Random ().Next () by default.  This is not the same value every time the program is run, even if it is reloaded.  new Random () "Initializes a new instance of the Random class, using a time-dependent default seed value. " 

So it appears that is not an issue.  If the random seed is set on the command line, then it is correctly set at the start of the operation to that value.

**Comment by stevenaw** [comment]
I'm curious if this is a duplicate of #1461 as both seem to seek making the seed generation more deterministic

**Comment by CharliePoole** [comment]
Essentially, this issue is about using the seed under a 3rd party, non-open-source runner, which does not reload the tests between runs. Back when I worked on the framework (i.e. 2017) I didn't take on the issue because it was specific to that environment and because @oznetmaster would eventually produce a PR to be reviewed. As it's 2023 now, someone may want to consider re-assigning or closing it. :-)

**Comment by CharliePoole** [comment]
Also, note that any runner is able to reload the tests, using the `RandomSeed` setting to select a specific seed if desired.
